### PR TITLE
Clean flow cell run directories

### PR DESCRIPTION
**Pull request by barrystokman**
## Description
This PR finds and removes flow cell run directories, and archives the sample sheets found in there before doing so.

### Added
Remove flow cell run directories and archive the sample sheets in Housekeeper bundles, based on sequencer type. Flow cells will be removed 21 days after sequencing by default, but any number of days can be specified.

### How to prepare for test
- [ ] ssh to `hasta`
- [ ] Use stage: `us`
- [ ] paxa the environment: `paxa`
- [ ] install on stage (example for hasta):
`bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-cg-stage.sh feat-clean-bcl-run-dirs`
- [ ] install on stage (example for hasta):
`bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-servers-stage.sh add-clean-run-dirs`

### How to test
Preconditions:
1. Prepare flow cell run directories:
    a.  Hiseqx:
`rsync -av /home/proj/production/flowcells/hiseqx/171215_ST-E00198_0265_AHFG22CCXY --exclude 'Data' --exclude '*bcl' --exclude '*tsv' --exclude 'InterOp' --exclude 'Images' --exclude 'Logs' /home/proj/stage/flowcells/hiseqx/`  
`rsync -av /home/proj/production/flowcells/hiseqx/181109_ST-E00201_0324_BHVCL7CCXY --exclude 'Data' --exclude '*bcl' --exclude '*tsv' --exclude 'InterOp' --exclude 'Images' --exclude 'Logs' /home/proj/stage/flowcells/hiseqx/`
    b. 2500:
`rsync -av /home/proj/production/flowcells/2500/runs/151127_SN7001301_0218_BHF277BCXX --exclude 'Data' --exclude '*bcl' --exclude 'InterOp' --exclude 'Images' --exclude 'Logs' /home/proj/stage/flowcells/2500/runs//`  
`rsync -av /home/proj/production/flowcells/2500/runs/210618_D00415_0973_AHKT22BCX3 --exclude 'Data' --exclude '*bcl' --exclude 'InterOp' --exclude 'Images' --exclude 'Logs' /home/proj/stage/flowcells/2500/runs//`
    c. NovaSeq:
`rsync -av /home/proj/production/flowcells/novaseq/runs/210204_A00689_0224_AHMWJNDSXY
 --exclude 'Data' --exclude '*bcl' --exclude 'InterOp' --exclude 'Images' --exclude 'Logs' /home/proj/stage/flowcells/novaseq/runs/`  
`rsync -av /home/proj/production/flowcells/novaseq/runs/220112_A00689_0450_BHNLNJDSX2 --exclude 'Data' --exclude '*bcl' --exclude 'InterOp' --exclude 'Images' --exclude 'Logs' /home/proj/stage/flowcells/novaseq/runs/`
`cp -R /home/proj/stage/flowcells/novaseq/runs/220112_A00689_0450_BHNLNJDSX2 /home/proj/stage/flowcells/novaseq/runs/220112_A00689_0450_BNOSHEET && rm /home/proj/stage/flowcells/novaseq/runs/220112_A00689_0450_BNOSHEET/SampleSheet.csv`
2. Prepare Housekeeper:
`for flow_cell in "HFG22CCXY" "HVCL7CCXY" "HF277BCXX" "HKT22BCX3" "HMWJNDSXY" "HNLNJDSX2" "NOSHEET"; do housekeeper delete version -b $flow_cell; housekeeper delete bundle $flow_cell; done`
3. Prepare `housekeeper-bundles`:
4. `for flow_cell in "HFG22CCXY" "HVCL7CCXY" "HF277BCXX" "HKT22BCX3" "HMWJNDSXY" "HNLNJDSX2" "NOSHEET"; do rm -rf /home/proj/stage/housekeeper-bundles/$flow_cell; done`

## Review
- [ ] code approved by
- [ ] tests executed by @barrystokman
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a
- [x] **MINOR** - when you add functionality in a backwards compatible manner

## Implementation Plan
- [ ] Deploy this branch
- [ ] Deploy https://github.com/Clinical-Genomics/servers/pull/733

### CODE REVIEW

**Comment by barrystokman** [comment]
TC1: help text
image

**Comment by barrystokman** [comment]
TC2: dry run, all sequencers, default age
image

**Comment by barrystokman** [comment]
TC3: dry run, select sequencer, non default age
hiseqx, 1200 days:
image
2500, 1200 days:
image
novaseq, 100 days
image

**Comment by barrystokman** [comment]
TC4: remove FCs and archive samples sheets, default input values (production like)
image
image
image

All flow cells older than 21 days are deleted:
image

Archived sample sheets are included in Housekeeper:
image

Note that flow cel 220112_A00689_0450_BHNLNJDSX2 and flow cell 220112_A00689_0450_BHNLNJDSX2 are not removed, and its sample sheets are not archived because the flow cells are not older than 21 days.

File: cg/meta/clean/flow_cell_run_directories.py
Comment (henrikstranneheim) [inline cg/meta/clean/flow_cell_run_directories.py:2]:
```suggestion
flow_cell_run_dir/<sequencer> """
```

File: cg/meta/clean/flow_cell_run_directories.py
Comment (henrikstranneheim) [inline cg/meta/clean/flow_cell_run_directories.py:63]:
```suggestion
```

File: cg/meta/clean/flow_cell_run_directories.py
Comment (henrikstranneheim) [inline cg/meta/clean/flow_cell_run_directories.py:91]: This code is awfully similiar to: 
```
   def add_sample_sheet_to_housekeeper(self, sample_sheet_path: Path):
        """Adds an archive sample sheet to Housekeeper"""

        hk_bundle: hk_models.Bundle = self.hk.bundle(self.id)
        if hk_bundle is None:
            hk_bundle = self.hk.create_new_bundle_and_version(name=self.id)

        with self.hk.session_no_autoflush():
            hk_version: hk_models.Version = self.hk.last_version(bundle=hk_bundle.name)
            if self.hk.files(path=str(sample_sheet_path)).first() is None:
                LOG.info(f"Adding archived sample sheet: {str(sample_sheet_path)}")
                tags: List[str] = [HousekeeperTags.ARCHIVED_SAMPLE_SHEET, self.id]
                self.hk.add_file(path=str(sample_sheet_path), version_obj=hk_version, tags=tags)
        self.hk.commit()
```
in cg/meta/clean/demultiplexed_flow_cells.py. Could this be move to HouseKeeperAPI to be used by both these classes. It seems unnecessary to repeat this code block in multiple places

File: tests/meta/clean/test_clean_flow_cell_run_directories.py
Comment (henrikstranneheim) [inline tests/meta/clean/test_clean_flow_cell_run_directories.py:73]:
```suggestion
    # GIVEN a sample sheet connected to the flow cell
```
I think is is more readable to stick to multiple GIVEN in the `#GIVEN-WHEN-THEN` instead of just comments

**Review by henrikstranneheim — COMMENTED** [review]
Nice!

File: tests/meta/clean/test_clean_flow_cell_run_directories.py
Comment (barrystokman) [reply tests/meta/clean/test_clean_flow_cell_run_directories.py]: fixed in https://github.com/Clinical-Genomics/cg/pull/1371/commits/b8ef51695f0716a0c8ae67b010f094150e8e20b0

File: cg/meta/clean/flow_cell_run_directories.py
Comment (barrystokman) [reply cg/meta/clean/flow_cell_run_directories.py]: I removed this entirely and used Housekeeper's `add_and_include_file_to_latest_version` method instead, which I somehow missed before.

in PR https://github.com/Clinical-Genomics/cg/pull/1334, it was initially agreed upon that the sample sheets should be archived in a separate directory, but now the requirement is to include those sample sheets instead. I can change that too, but I'm reluctant to do that in this PR because it will require time consuming retesting on stage.

Comment (henrikstranneheim) [reply cg/meta/clean/flow_cell_run_directories.py]: Agreed!

**Comment by barrystokman** [comment]
Retest:
image
image
image

**Comment by barrystokman** [comment]
Ready to merge and deploy after I get an approval.

File: cg/meta/clean/flow_cell_run_directories.py
Comment (henrikstranneheim) [inline cg/meta/clean/flow_cell_run_directories.py:24]:
```suggestion
    """Class to check and remove flow cell run directories"""
```
Less is more. I do not think this class controls the 21 days.

File: cg/meta/clean/flow_cell_run_directories.py
Comment (henrikstranneheim) [inline cg/meta/clean/flow_cell_run_directories.py:46]:
```suggestion
            self._age: int = (datetime.now() - self.sequenced_date).days
```

File: cg/meta/clean/flow_cell_run_directories.py
Comment (henrikstranneheim) [inline cg/meta/clean/flow_cell_run_directories.py:56]:
```suggestion
                self._sequenced_date: datetime = self.status_db.flowcell(name=self.id).sequenced_at
```

File: cg/meta/clean/flow_cell_run_directories.py
Comment (henrikstranneheim) [inline cg/meta/clean/flow_cell_run_directories.py:62]:
```suggestion
                self._sequenced_date: datetime = self.derived_date
```

File: cg/meta/clean/flow_cell_run_directories.py
Comment (henrikstranneheim) [inline cg/meta/clean/flow_cell_run_directories.py:88]: ⭐ So much nicer and more maintainable!

**Review by henrikstranneheim — APPROVED** [review]
Excellent!

**Comment by barrystokman** [comment]
image

**Pull request merged**
